File: PROYECTO_RESTFULL_JAVA_BDD/02 CLIESC/ComercializadoraEscritorio/src/main/java/ec/edu/monster/services/ClienteUnificado.java
package ec.edu.monster.services;

import ec.edu.monster.models.*;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

/**
 * Cliente unificado con soporte completo para REST (Java) y SOAP (.NET)
 * @author josue
 */
public class ClienteUnificado {

    public enum TipoProtocolo {
        REST, SOAP
    }

    private final HttpClient httpClient;
    private final Gson gson;
    private TipoProtocolo protocoloActual;
    private final SoapClient soapClient;

    // Endpoints REST (Java)
    private static final String BASE_COMERCIALIZADORA = "http://localhost:8080/ComercializadoraElectrodomesticos/api";
    private static final String BASE_BANQUITO = "http://localhost:8080/BanquitoCore/api/credito";

    public ClienteUnificado() {
        this.httpClient = HttpClient.newBuilder()
                .build();

        this.gson = new GsonBuilder()
                .setPrettyPrinting()
                .create();

        this.soapClient = new SoapClient();
        this.protocoloActual = TipoProtocolo.REST;
    }

    public CompletableFuture<List<Factura>> listarFacturasAsync() {
        return listarFacturasREST();
    }

    public CompletableFuture<RespuestaFactura> obtenerFacturaAsync(int idFactura) {
        return obtenerFacturaREST(idFactura);
    }

    // ========== CONFIGURACIÓN ==========

    public TipoProtocolo getProtocoloActual() {
        return protocoloActual;
    }

    public void cambiarProtocolo(TipoProtocolo nuevoProtocolo) {
        this.protocoloActual = nuevoProtocolo;
    }

    // ========== ELECTRODOMÉSTICOS ==========

    public CompletableFuture<List<Electrodomestico>> listarElectrodomesticosAsync() {
        if (protocoloActual == TipoProtocolo.REST) {
            return listarElectrodomesticosREST();
        } else {
            return soapClient.listarElectrodomesticosSOAP();
        }
    }

    public CompletableFuture<Electrodomestico> obtenerElectrodomesticoAsync(int id) {
        if (protocoloActual == TipoProtocolo.REST) {
            return obtenerElectrodomesticoREST(id);
        } else {
            return soapClient.obtenerElectrodomesticoSOAP(id);
        }
    }

    public CompletableFuture<RespuestaOperacion> crearElectrodomesticoAsync(Electrodomestico producto) {
        if (protocoloActual == TipoProtocolo.REST) {
            return crearElectrodomesticoREST(producto);
        } else {
            return soapClient.crearElectrodomesticoSOAP(producto);
        }
    }

    public CompletableFuture<RespuestaOperacion> actualizarElectrodomesticoAsync(int id, Electrodomestico producto) {
        if (protocoloActual == TipoProtocolo.REST) {
            return actualizarElectrodomesticoREST(id, producto);
        } else {
            return soapClient.actualizarElectrodomesticoSOAP(id, producto);
        }
    }

    public CompletableFuture<RespuestaOperacion> eliminarElectrodomesticoAsync(int id) {
        if (protocoloActual == TipoProtocolo.REST) {
            return eliminarElectrodomesticoREST(id);
        } else {
            return soapClient.eliminarElectrodomesticoSOAP(id);
        }
    }

    // ========== FACTURACIÓN ==========

    public CompletableFuture<RespuestaVenta> procesarVentaEfectivoAsync(SolicitudVenta solicitud) {
        if (protocoloActual == TipoProtocolo.REST) {
            return procesarVentaEfectivoREST(solicitud);
        } else {
            return soapClient.procesarVentaEfectivoSOAP(solicitud);
        }
    }

    public CompletableFuture<RespuestaVenta> procesarVentaCreditoAsync(SolicitudVenta solicitud) {
        if (protocoloActual == TipoProtocolo.REST) {
            return procesarVentaCreditoREST(solicitud);
        } else {
            return soapClient.procesarVentaCreditoSOAP(solicitud);
        }
    }

    // ========== CRÉDITO BANQUITO ==========

    public CompletableFuture<ValidacionCredito> validarSujetoCreditoAsync(String cedula) {
        if (protocoloActual == TipoProtocolo.REST) {
            return validarSujetoCreditoREST(cedula);
        } else {
            return soapClient.validarSujetoCreditoSOAP(cedula);
        }
    }

    public CompletableFuture<MontoMaximo> obtenerMontoMaximoAsync(String cedula) {
        if (protocoloActual == TipoProtocolo.REST) {
            return obtenerMontoMaximoREST(cedula);
        } else {
            return soapClient.obtenerMontoMaximoSOAP(cedula);
        }
    }

    public CompletableFuture<TablaAmortizacion> obtenerTablaAmortizacionAsync(int idCredito) {
        if (protocoloActual == TipoProtocolo.REST) {
            return obtenerTablaAmortizacionREST(idCredito);
        } else {
            return soapClient.obtenerTablaAmortizacionSOAP(idCredito);
        }
    }

    // ========== CONECTIVIDAD ==========

    public CompletableFuture<Boolean> probarConectividadAsync() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                if (protocoloActual == TipoProtocolo.REST) {
                    var testBanquito = testBanquitoREST().get();
                    var testComercializadora = listarElectrodomesticosREST().get();
                    return testBanquito != null && testComercializadora != null;
                } else {
                    // Para SOAP, probar conexión básica
                    var testComercializadora = soapClient.listarElectrodomesticosSOAP().get();
                    var testBanquito = soapClient.validarSujetoCreditoSOAP("1234567890").get();
                    return testComercializadora != null && testBanquito != null;
                }
            } catch (Exception e) {
                return false;
            }
        });
    }

    // ========== IMPLEMENTACIONES REST ==========

    private CompletableFuture<List<Electrodomestico>> listarElectrodomesticosREST() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/electrodomesticos"))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(),
                            new TypeToken<List<Electrodomestico>>(){}.getType());
                }
                return null;
            } catch (Exception e) {
                throw new RuntimeException("Error al listar electrodomésticos: " + e.getMessage(), e);
            }
        });
    }

    private CompletableFuture<Electrodomestico> obtenerElectrodomesticoREST(int id) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/electrodomesticos/" + id))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(), Electrodomestico.class);
                }
                return null;
            } catch (Exception e) {
                throw new RuntimeException("Error al obtener electrodoméstico: " + e.getMessage(), e);
            }
        });
    }

    private CompletableFuture<RespuestaOperacion> crearElectrodomesticoREST(Electrodomestico producto) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String jsonBody = gson.toJson(producto);

                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/electrodomesticos"))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                RespuestaOperacion respuesta = gson.fromJson(response.body(), RespuestaOperacion.class);
                return respuesta != null ? respuesta : new RespuestaOperacion(false, "Error desconocido");

            } catch (Exception e) {
                return new RespuestaOperacion(false, "Error al crear producto: " + e.getMessage());
            }
        });
    }

    private CompletableFuture<RespuestaOperacion> actualizarElectrodomesticoREST(int id, Electrodomestico producto) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String jsonBody = gson.toJson(producto);

                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/electrodomesticos/" + id))
                        .header("Content-Type", "application/json")
                        .PUT(HttpRequest.BodyPublishers.ofString(jsonBody))
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                RespuestaOperacion respuesta = gson.fromJson(response.body(), RespuestaOperacion.class);
                return respuesta != null ? respuesta : new RespuestaOperacion(false, "Error desconocido");

            } catch (Exception e) {
                return new RespuestaOperacion(false, "Error al actualizar producto: " + e.getMessage());
            }
        });
    }

    private CompletableFuture<RespuestaOperacion> eliminarElectrodomesticoREST(int id) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/electrodomesticos/" + id))
                        .DELETE()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                RespuestaOperacion respuesta = gson.fromJson(response.body(), RespuestaOperacion.class);
                return respuesta != null ? respuesta : new RespuestaOperacion(false, "Error desconocido");

            } catch (Exception e) {
                return new RespuestaOperacion(false, "Error al eliminar producto: " + e.getMessage());
            }
        });
    }

    private CompletableFuture<RespuestaVenta> procesarVentaEfectivoREST(SolicitudVenta solicitud) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String jsonBody = gson.toJson(solicitud);

                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/facturacion/venta-efectivo"))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                return gson.fromJson(response.body(), RespuestaVenta.class);

            } catch (Exception e) {
                RespuestaVenta error = new RespuestaVenta();
                error.setExito(false);
                error.setMensaje("Error al procesar venta: " + e.getMessage());
                return error;
            }
        });
    }

    private CompletableFuture<RespuestaVenta> procesarVentaCreditoREST(SolicitudVenta solicitud) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                String jsonBody = gson.toJson(solicitud);

                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/facturacion/venta-credito"))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                return gson.fromJson(response.body(), RespuestaVenta.class);

            } catch (Exception e) {
                RespuestaVenta error = new RespuestaVenta();
                error.setExito(false);
                error.setMensaje("Error al procesar venta a crédito: " + e.getMessage());
                return error;
            }
        });
    }

    private CompletableFuture<ValidacionCredito> validarSujetoCreditoREST(String cedula) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_BANQUITO + "/validar/" + cedula))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(), ValidacionCredito.class);
                }

                ValidacionCredito error = new ValidacionCredito();
                error.setSujetoCredito(false);
                error.setMensaje("Error en el servicio de validación");
                return error;

            } catch (Exception e) {
                ValidacionCredito error = new ValidacionCredito();
                error.setSujetoCredito(false);
                error.setMensaje("Error de conexión: " + e.getMessage());
                return error;
            }
        });
    }

    private CompletableFuture<MontoMaximo> obtenerMontoMaximoREST(String cedula) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_BANQUITO + "/monto-maximo/" + cedula))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(), MontoMaximo.class);
                }

                MontoMaximo error = new MontoMaximo();
                error.setAprobado(false);
                error.setMensaje("Error en el servicio");
                return error;

            } catch (Exception e) {
                MontoMaximo error = new MontoMaximo();
                error.setAprobado(false);
                error.setMensaje("Error de conexión: " + e.getMessage());
                return error;
            }
        });
    }

    private CompletableFuture<TablaAmortizacion> obtenerTablaAmortizacionREST(int idCredito) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_BANQUITO + "/tabla-amortizacion/" + idCredito))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(), TablaAmortizacion.class);
                }

                TablaAmortizacion error = new TablaAmortizacion();
                error.setEncontrado(false);
                error.setMensaje("Tabla de amortización no encontrada");
                return error;

            } catch (Exception e) {
                TablaAmortizacion error = new TablaAmortizacion();
                error.setEncontrado(false);
                error.setMensaje("Error de conexión: " + e.getMessage());
                return error;
            }
        });
    }

    private CompletableFuture<TestConectividad> testBanquitoREST() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_BANQUITO + "/test"))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(), TestConectividad.class);
                }
                return null;

            } catch (Exception e) {
                return null;
            }
        });
    }

    private CompletableFuture<List<Factura>> listarFacturasREST() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/facturacion/facturas"))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    return gson.fromJson(response.body(),
                            new TypeToken<List<Factura>>(){}.getType());
                }
                return new ArrayList<>();
            } catch (Exception e) {
                throw new RuntimeException("Error al listar facturas: " + e.getMessage(), e);
            }
        });
    }

    private CompletableFuture<RespuestaFactura> obtenerFacturaREST(int idFactura) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(URI.create(BASE_COMERCIALIZADORA + "/facturacion/facturas/" + idFactura))
                        .GET()
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() == 200) {
                    // La respuesta contiene la factura directamente con algunos campos adicionales
                    Factura factura = gson.fromJson(response.body(), Factura.class);

                    RespuestaFactura respuesta = new RespuestaFactura();
                    respuesta.setEncontrada(true);
                    respuesta.setMensaje("Factura obtenida exitosamente");
                    respuesta.setFactura(factura);

                    return respuesta;
                } else {
                    RespuestaFactura respuesta = new RespuestaFactura();
                    respuesta.setEncontrada(false);
                    respuesta.setMensaje("Factura no encontrada");
                    return respuesta;
                }
            } catch (Exception e) {
                RespuestaFactura respuesta = new RespuestaFactura();
                respuesta.setEncontrada(false);
                respuesta.setMensaje("Error al obtener factura: " + e.getMessage());
                return respuesta;
            }
        });
    }
}

File: PROYECTO_RESTFULL_JAVA_BDD/02 CLIMOV/ClienteMovil/ClienteMovil/Services/ClienteUnificado.cs
using ClienteMovil.Models;
using System.Text.Json;

namespace ClienteMovil.Services
{
    /// <summary>
    /// Cliente unificado para consumir servicios REST y SOAP
    /// </summary>
    public class ClienteUnificado
    {
        private readonly HttpClient _httpClient;
        private readonly JsonSerializerOptions _jsonOptions;

        public TipoProtocolo ProtocoloActual { get; private set; }

        public ClienteUnificado()
        {
            _httpClient = new HttpClient
            {
                Timeout = ConfiguracionTimeouts.TIMEOUT_DEFAULT
            };

            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true,
                WriteIndented = true
            };

            // Por defecto usar REST
            ProtocoloActual = TipoProtocolo.REST;
        }

        public void CambiarProtocolo(TipoProtocolo nuevoProtocolo)
        {
            ProtocoloActual = nuevoProtocolo;
        }

        // ========== ELECTRODOMÉSTICOS ==========

        public async Task<List<Electrodomestico>?> ListarElectrodomesticosAsync()
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ListarElectrodomesticosREST();
                }
                else
                {
                    // return await ListarElectrodomesticosSOAP();
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error en ListarElectrodomesticos: {ex.Message}");
                return null;
            }
        }

        public async Task<Electrodomestico?> ObtenerElectrodomesticoAsync(int id)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ObtenerElectrodomesticoREST(id);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error en ObtenerElectrodomestico: {ex.Message}");
                return null;
            }
        }

        public async Task<RespuestaOperacion?> CrearElectrodomesticoAsync(Electrodomestico electrodomestico)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await CrearElectrodomesticoREST(electrodomestico);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new RespuestaOperacion
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        public async Task<RespuestaOperacion?> ActualizarElectrodomesticoAsync(int id, Electrodomestico electrodomestico)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ActualizarElectrodomesticoREST(id, electrodomestico);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new RespuestaOperacion
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        public async Task<RespuestaOperacion?> EliminarElectrodomesticoAsync(int id)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await EliminarElectrodomesticoREST(id);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new RespuestaOperacion
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        // ========== FACTURACIÓN ==========

        public async Task<RespuestaVenta?> ProcesarVentaEfectivoAsync(SolicitudVenta solicitud)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ProcesarVentaEfectivoREST(solicitud);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new RespuestaVenta
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        public async Task<RespuestaVenta?> ProcesarVentaCreditoAsync(SolicitudVenta solicitud)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ProcesarVentaCreditoREST(solicitud);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new RespuestaVenta
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        // ========== CRÉDITO BANQUITO ==========

        public async Task<ValidacionCredito?> ValidarSujetoCreditoAsync(string cedula)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ValidarSujetoCreditoREST(cedula);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new ValidacionCredito
                {
                    SujetoCredito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        public async Task<MontoMaximo?> ObtenerMontoMaximoAsync(string cedula)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ObtenerMontoMaximoREST(cedula);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new MontoMaximo
                {
                    Aprobado = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        public async Task<TablaAmortizacion?> ObtenerTablaAmortizacionAsync(int idCredito)
        {
            try
            {
                if (ProtocoloActual == TipoProtocolo.REST)
                {
                    return await ObtenerTablaAmortizacionREST(idCredito);
                }
                else
                {
                    throw new NotImplementedException("SOAP será implementado en la siguiente fase");
                }
            }
            catch (Exception ex)
            {
                return new TablaAmortizacion
                {
                    Encontrado = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        // ========== CONECTIVIDAD ==========

        public async Task<EstadoConectividad> ProbarConectividadAsync()
        {
            var estado = new EstadoConectividad
            {
                ProtocoloActual = ProtocoloActual
            };

            try
            {
                // Probar Comercializadora
                var productos = await ListarElectrodomesticosAsync();
                estado.ComercializadoraActiva = productos != null;

                // Probar BanQuito
                var test = await ProbarConectividadBanQuito();
                estado.BanquitoActivo = test?.EsActivo == true;
            }
            catch
            {
                estado.ComercializadoraActiva = false;
                estado.BanquitoActivo = false;
            }

            return estado;
        }

        // ========== IMPLEMENTACIONES REST ==========

        private async Task<List<Electrodomestico>?> ListarElectrodomesticosREST()
        {
            try
            {
                var response = await _httpClient.GetAsync(ConfiguracionEndpoints.REST.ELECTRODOMESTICOS);

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<List<Electrodomestico>>(json, _jsonOptions);
                }

                return null;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error en ListarElectrodomesticosREST: {ex.Message}");
                return null;
            }
        }

        private async Task<Electrodomestico?> ObtenerElectrodomesticoREST(int id)
        {
            try
            {
                var response = await _httpClient.GetAsync(ConfiguracionEndpoints.REST.ElectrodomesticoPorId(id));

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<Electrodomestico>(json, _jsonOptions);
                }

                return null;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error en ObtenerElectrodomesticoREST: {ex.Message}");
                return null;
            }
        }

        private async Task<RespuestaOperacion?> CrearElectrodomesticoREST(Electrodomestico electrodomestico)
        {
            try
            {
                var json = JsonSerializer.Serialize(electrodomestico, _jsonOptions);
                var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");

                var response = await _httpClient.PostAsync(ConfiguracionEndpoints.REST.ELECTRODOMESTICOS, content);

                if (response.IsSuccessStatusCode)
                {
                    var responseJson = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<RespuestaOperacion>(responseJson, _jsonOptions);
                }
                else
                {
                    return new RespuestaOperacion
                    {
                        Exito = false,
                        Mensaje = $"Error HTTP: {response.StatusCode}"
                    };
                }
            }
            catch (Exception ex)
            {
                return new RespuestaOperacion
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<RespuestaOperacion?> ActualizarElectrodomesticoREST(int id, Electrodomestico electrodomestico)
        {
            try
            {
                var json = JsonSerializer.Serialize(electrodomestico, _jsonOptions);
                var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");

                var response = await _httpClient.PutAsync(ConfiguracionEndpoints.REST.ElectrodomesticoPorId(id), content);

                if (response.IsSuccessStatusCode)
                {
                    var responseJson = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<RespuestaOperacion>(responseJson, _jsonOptions);
                }
                else
                {
                    return new RespuestaOperacion
                    {
                        Exito = false,
                        Mensaje = $"Error HTTP: {response.StatusCode}"
                    };
                }
            }
            catch (Exception ex)
            {
                return new RespuestaOperacion
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<RespuestaOperacion?> EliminarElectrodomesticoREST(int id)
        {
            try
            {
                var response = await _httpClient.DeleteAsync(ConfiguracionEndpoints.REST.ElectrodomesticoPorId(id));

                if (response.IsSuccessStatusCode)
                {
                    var responseJson = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<RespuestaOperacion>(responseJson, _jsonOptions);
                }
                else
                {
                    return new RespuestaOperacion
                    {
                        Exito = false,
                        Mensaje = $"Error HTTP: {response.StatusCode}"
                    };
                }
            }
            catch (Exception ex)
            {
                return new RespuestaOperacion
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<RespuestaVenta?> ProcesarVentaEfectivoREST(SolicitudVenta solicitud)
        {
            try
            {
                var json = JsonSerializer.Serialize(solicitud, _jsonOptions);
                var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");

                var response = await _httpClient.PostAsync(ConfiguracionEndpoints.REST.VENTA_EFECTIVO, content);
                var responseJson = await response.Content.ReadAsStringAsync();

                return JsonSerializer.Deserialize<RespuestaVenta>(responseJson, _jsonOptions);
            }
            catch (Exception ex)
            {
                return new RespuestaVenta
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<RespuestaVenta?> ProcesarVentaCreditoREST(SolicitudVenta solicitud)
        {
            try
            {
                var json = JsonSerializer.Serialize(solicitud, _jsonOptions);
                var content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");

                var response = await _httpClient.PostAsync(ConfiguracionEndpoints.REST.VENTA_CREDITO, content);
                var responseJson = await response.Content.ReadAsStringAsync();

                return JsonSerializer.Deserialize<RespuestaVenta>(responseJson, _jsonOptions);
            }
            catch (Exception ex)
            {
                return new RespuestaVenta
                {
                    Exito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<ValidacionCredito?> ValidarSujetoCreditoREST(string cedula)
        {
            try
            {
                var response = await _httpClient.GetAsync(ConfiguracionEndpoints.REST.ValidarCredito(cedula));

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<ValidacionCredito>(json, _jsonOptions);
                }

                return new ValidacionCredito
                {
                    SujetoCredito = false,
                    Mensaje = "Error en el servicio"
                };
            }
            catch (Exception ex)
            {
                return new ValidacionCredito
                {
                    SujetoCredito = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<MontoMaximo?> ObtenerMontoMaximoREST(string cedula)
        {
            try
            {
                var response = await _httpClient.GetAsync(ConfiguracionEndpoints.REST.MontoMaximo(cedula));

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<MontoMaximo>(json, _jsonOptions);
                }

                return new MontoMaximo
                {
                    Aprobado = false,
                    Mensaje = "Error en el servicio"
                };
            }
            catch (Exception ex)
            {
                return new MontoMaximo
                {
                    Aprobado = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<TablaAmortizacion?> ObtenerTablaAmortizacionREST(int idCredito)
        {
            try
            {
                var response = await _httpClient.GetAsync(ConfiguracionEndpoints.REST.TablaAmortizacion(idCredito));

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<TablaAmortizacion>(json, _jsonOptions);
                }

                return new TablaAmortizacion
                {
                    Encontrado = false,
                    Mensaje = "Tabla no encontrada"
                };
            }
            catch (Exception ex)
            {
                return new TablaAmortizacion
                {
                    Encontrado = false,
                    Mensaje = $"Error: {ex.Message}"
                };
            }
        }

        private async Task<TestConectividad?> ProbarConectividadBanQuito()
        {
            try
            {
                var response = await _httpClient.GetAsync(ConfiguracionEndpoints.REST.TEST_BANQUITO);

                if (response.IsSuccessStatusCode)
                {
                    var json = await response.Content.ReadAsStringAsync();
                    return JsonSerializer.Deserialize<TestConectividad>(json, _jsonOptions);
                }
                return null;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error en ProbarConectividadBanQuito: {ex.Message}");
                return null;
            }
        }

        // ========== CLEANUP ==========

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }
}
